File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h2x/low_level/low_state.cpp
/*
 * 注意事项
 *
 * 1. 机器人模式切换需要确保当前状态安全
 * 2. 移动命令需要在合适的模式下执行
 * 3. 所有操作结果通过result字段返回，0表示成功，非0表示错误
 */

// low_state.cpp

#include <iostream>
#include <thread>
#include <chrono>

#include "common/common_def.h"
#include "common/h2x_def.h"
#include "h2x/h2x_low_level.h"

using namespace humanoid;
using namespace humanoid::h2x;

int main() {
  std::cout << "=== Low State Reading Test ===" << std::endl;

  H2xLowLevel low_level;
  LowState low_state;

  // Read low state for 5 seconds
  constexpr int DURATION_MS = 5000;
  constexpr int READ_INTERVAL_MS = 100;
  constexpr int read_count = DURATION_MS / READ_INTERVAL_MS;

  for (int i = 0; i < read_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();

    int ret = low_level.readLowState(low_state);
    if (ret != 0) {
      std::cout << "Failed to read low state!" << std::endl;
      return -1;
    }

    // Print some state information
    std::cout << "=== Low State Reading " << i + 1 << " ===" << std::endl;
    std::cout << "Tick: " << low_state.tick << std::endl;
    std::cout << "Time seconds: " << low_state.time_seconds << std::endl;
    std::cout << "Is data valid: " << (low_state.is_data_valid ? "Yes" : "No")
              << std::endl;

    // Print first motor state as example
    if (low_state.is_data_valid) {
      std::cout << "Motor 0 - Position: " << low_state.motor_state[0].q
                << ", Velocity: " << low_state.motor_state[0].dq
                << ", Torque: " << low_state.motor_state[0].tau_est
                << std::endl;
    }

    // Print IMU state if available
    if (low_state.is_data_valid) {
      std::cout << "IMU 0 - Accel: [" << low_state.imu_state[0].accelerometer[0]
                << ", " << low_state.imu_state[0].accelerometer[1] << ", "
                << low_state.imu_state[0].accelerometer[2] << "]"
                << ", Gyro: [" << low_state.imu_state[0].gyroscope[0] << ", "
                << low_state.imu_state[0].gyroscope[1] << ", "
                << low_state.imu_state[0].gyroscope[2] << "]" << std::endl;
    }

    std::cout << "Error code: " << low_state.error_code << std::endl;
    std::cout << "Warning code: " << low_state.warning_code << std::endl;
    std::cout << std::endl;

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed =
        std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::this_thread::sleep_for(std::chrono::milliseconds(READ_INTERVAL_MS) -
                                elapsed);
  }

  std::cout << "Low state reading test completed successfully" << std::endl;
  return 0;
}

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h2x/low_level/low_state.py
#!/usr/bin/env python3
#!
#!注意事项
#!
#!1. 机器人模式切换需要确保当前状态安全
#!2. 移动命令需要在合适的模式下执行
#!3. 所有操作结果通过result字段返回，0表示成功，非0表示错误


import humanoid_sdk_py
from humanoid_sdk_py import h2x
import time


def main():
    print("=== Low State Reading Test ===")

    # Initialize low level object
    low_level = h2x.H2xLowLevel()

    # Read low state for 5 seconds
    DURATION_MS = 5000
    READ_INTERVAL_MS = 100
    read_count = DURATION_MS // READ_INTERVAL_MS

    for i in range(read_count):
        start_time = time.time()

        # Read low state
        ret, low_state = low_level.readLowState()
        if ret != 0:
            print("Failed to read low state!")
            return -1

        # Print some state information
        print(f"=== Low State Reading {i + 1} ===")
        print(f"Tick: {low_state.tick}")
        print(f"Time seconds: {low_state.time_seconds}")
        print(f"Is data valid: {'Yes' if low_state.is_data_valid else 'No'}")
        
        # Print first motor state as example
        print(f"Motor 0 - Position: {low_state.motor_state[0].q:.4f}, "
              f"Velocity: {low_state.motor_state[0].dq:.4f}, "
              f"Torque: {low_state.motor_state[0].tau_est:.4f}")

        # Print IMU state if available
        imu = low_state.imu_state[0]
        print(f"IMU 0 - Accel: [{imu.accelerometer[0]:.4f}, "
              f"{imu.accelerometer[1]:.4f}, {imu.accelerometer[2]:.4f}], "
              f"Gyro: [{imu.gyroscope[0]:.4f}, "
              f"{imu.gyroscope[1]:.4f}, {imu.gyroscope[2]:.4f}]")

        print(f"Error code: {low_state.error_code}")
        print(f"Warning code: {low_state.warning_code}")
        print()

        elapsed_time = (time.time() - start_time) * 1000
        sleep_time = READ_INTERVAL_MS - elapsed_time
        if sleep_time > 0:
            time.sleep(sleep_time / 1000.0)

    print("Low state reading test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())